api/hypershift/v1beta1/certificaterevocationrequest_types.go

```diff
@@ -0,0 +1,96 @@
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	SchemeBuilder.Register(&CertificateRevocationRequest{})
+	SchemeBuilder.Register(&CertificateRevocationRequestList{})
+}
+
+// +genclient
+// +kubebuilder:resource:path=certificaterevocationrequest,shortName=crr;crrs,scope=Namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+
+// CertificateRevocationRequest defines the desired state of CertificateRevocationRequest
+type CertificateRevocationRequest struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   CertificateRevocationRequestSpec   `json:"spec,omitempty"`
+	Status CertificateRevocationRequestStatus `json:"status,omitempty"`
+}
+
+// CertificateRevocationRequestSpec defines the desired state of CertificateRevocationRequest
+type CertificateRevocationRequestSpec struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=customer-break-glass
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="signerClass is immutable"
+
+	// SignerClass identifies the class of signer to revoke. All the active signing CAs for the
+	// signer class will be revoked.
+	SignerClass string `json:"signerClass"`
+}
+
+type CertificateRevocationRequestPhase string
+
+const (
+	CertificateRevocationRequestPhaseUnknown      CertificateRevocationRequestPhase = "Unknown"
```

Comment:
Phase state machines were replaced by conditions.  Conditions have machine readable Reasons if you'd like to have a `Revoked` condition that is `False` with a reason of one of these phases.  Conditions make it easier to have code grow additional logical states without breaking API consumers.

Reply:
reading through, having independent conditions seems ideal so we don't lose the information as we go and info (both historical current) is easy to get for both machines and humans

Reply:
Is there an object you'd consider a good example of the new paradigm?

Reply:
Pods.

I'd say that each phase here could be a condition so we track and don't lose the info as we go through.

Reply:
As far as I can tell Pods still use `.status.phase` for a state machine - sorry if I'm being dense but what kinds of questions are we trying to ask such that we'd want to add a bunch to conditions? FWIW we emit events at every step of this process.

Reply:
The phases as written are necessary for the internal state machine to function - we need to store that state somewhere. Are you suggesting that it can't be in status unless clients have a need to know?

Reply:
> we need to store that state somewhere
I'd argue that the current state in your state machine (apart from maybe the revocation timestamp) can be derived from other resources in the namespace and doesn't need to be stored explicitly in the status.